proposals/NNNN-noncopyable-generics.md

```diff
@@ -0,0 +1,975 @@
+> `Copyable` type from being substituted for `T`, i.e., it does _not_ narrow the
+> kinds of types permitted, like a constraint would. This is the reason why
+> `~Copyable` is referred to as the _inverse_ of `Copyable` rather than a
+> constraint itself. When applying an inverse `~C` to a type variable implicitly
+> constrained to `C`, that implicit constraint is nullified. Removing any kind
+> of constraint from a generic parameter _broadens_ the kinds of types
+> permitted.
+
+As with a concrete noncopyable type, any generic type parameter that does not
+conform `Copyable` must use one of the ownership modifiers `borrowing`,
+`consuming`, or `inout`, when it appears as the type of a function's parameter.
+For details on these parameter ownership modifiers, see
+[SE-377](0377-parameter-ownership-modifiers.md).
+
+### Structs and enums
+Struct and enum types default to implicitly conforming to `Copyable`:
+
+```swift
+struct DataSet /* : Copyable */ {
+  var samples: [Double]
+}
+```
+
+As with any other generic parameter, there is a `Copyable` constraint on the
+`Elm` in this `List` too:
+
+```swift
+enum List<Elm> /* : Copyable where Elm: Copyable */ {
+  case empty
+  indirect case elm(Elm, List<Elm>)
+}
+```
+
+As in SE-390, a generic struct or enum can use the inverse `~Copyable` to
+remove the implicit `Copyable` conformance and constraint:
+
+```swift
+enum List<Elm: ~Copyable>: ~Copyable { /* ... */ }
+```
+
+Once `Elm` does not have a `Copyable` constraint, a noncopyable type like
+`FileDescriptor` can be substituted in `List`, in addition to copyable ones.
+
+#### Automatic synthesis of conditional `Copyable` conformance
+A common reason for having a type parameter is that values of that generic type
+will be stored somewhere within the parameterized struct or enum type. Because
+of `Copyable`'s containment rule, that implies the parameterized type itself
+is commonly noncopyable if any of its type parameters is noncopyable. This
+observation is one of the guiding principles behind the rules in this section.
+
+Using the inverse `~Copyable` on any type parameter of a struct or enum, without
+specifying the copying behavior for the parameterized type, will automatically
+synthesize a conditional conformance to `Copyable`:
+
+```swift
+// A conditionally-copyable pair.
+struct Pair<Elm: ~Copyable> {
+  // signature <Elm>
+  func exchangeFirst(_ new: consuming Elm) -> Elm? { /* ... */ }
+}
+/* extension Pair: Copyable where Elm: Copyable {} */
+
+// Pair<FileDescriptor> is NOT Copyable
+// Pair<Int> is Copyable
+
+// A conditionally-copyable factory.
+struct Factory<Product: ~Copyable, Token: ~Copyable, Policy> {
+  var seed = 0
+
+  // signature <Product, Token, Policy where Policy : Copyable>
+  mutating func produce(_ payment: consuming Token) -> Product { /* ... */ }
+}
+/* extension Factory: Copyable where Product: Copyable, Token: Copyable {} */
+```
+
+Notice that despite this `Factory` not storing any noncopyable data, it does not
```

Comment:
This might need justification, about how you wouldn't want the synthesizing behavior to have to be determined by the implementation but by the signature.

Reply:
One issue with the justification is that Sendable does it differently; it won't synthesize the Sendable implicitly if some storage is not Sendable. Will need to dig into the details and see if we can get them to work the same.